File: hyejee/네트워크.swift
//
//  main.swift
//  programmers-swift
//
//  Created by HyeeJee Kim on 2022/04/09.
//

import Foundation

func dfs(computer: Int, computers: [[Int]], visited: inout [Bool]) {
    visited[computer] = true
    
    for index in 0..<computers.count {
        if computers[computer][index] == 0 {
            continue
        }
        
        if visited[index] {
            continue
        }
        
        dfs(computer: index, computers: computers, visited: &visited)
    }
}

func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    var visited: [Bool] = Array(repeating: false, count: n)
    var networkCount: Int = 0
    
    // 방문하지 않은 컴퓨터를 시작 노드로 삼아 네트워크 찾기 시작
    for computer in 0..<n {
        if visited[computer] {
            continue
        }
        
        dfs(computer: computer, computers: computers, visited: &visited)
        networkCount += 1
    }
    
    return networkCount
}

File: chaewon/week8/네트워크.py
def dfs(n, computers, computer, visited):
    visited[computer] = True
    for other_computer in range(n):
        if other_computer != computer and computers[computer][other_computer] == 1:
            if not visited[other_computer]:
                dfs(n, computers, other_computer, visited)


def solution(n, computers):
    answer = 0
    visited = [False for _ in range(n)]
    for computer in range(n):
        if not visited[computer]:
            dfs(n, computers, computer, visited)
            answer += 1

    return answer